### PR TITLE
Proper rocket Holding

### PR DESCRIPTION
**Pull request by aronwk-aaron**
Implement proper rocket holding at launchpads.

Tested all world launchers with and without rocket.
Tested all property launchers with and without rocket.
Tested LUP launcher with and without rocket.
Tested double clicking or dragging rocket on to each type of launcher as well and interacting with them through <shift>.

In the scenarios above, when the player has a rocket, the launcher works as expected.
World launcher launch immediately
Prop and LUP launchers open then menu and waits, then launch when a selection is made.

When the player has no rocket, no launch sequence happen and no menus are opened.

Tested quitting the game/killing the server while player is holding rocket. Upon next login, player is not stuck holding rocket.

### CODE REVIEW

File: dGame/dComponents/CharacterComponent.cpp [hunk 1]
```diff
@@ -448,6 +449,79 @@ void CharacterComponent::SetLastRocketConfig(std::u16string config) {
...
+	}
+
+	if (!rocket) {
+		rocket = inventoryComponent->FindItemByLot(6416);
+	}
+
+	if (!rocket) {
+		Game::logger->Log("CharacterComponent", "Unable to find rocket to equip!\n");
+		return rocket;
+	}
```
Comment (EmosewaMC) [inline dGame/dComponents/CharacterComponent.cpp:471]: Should split this into its own method as it is repeated down below

File: dGame/dComponents/CharacterComponent.cpp [hunk 2]
```diff
@@ -448,6 +449,79 @@ void CharacterComponent::SetLastRocketConfig(std::u16string config) {
...
+
+	// build and defin the rocket config
```
Comment (EmosewaMC) [inline dGame/dComponents/CharacterComponent.cpp:473]: defin?

File: dGame/dComponents/PropertyEntranceComponent.cpp [hunk 3]
```diff
@@ -24,6 +25,15 @@ PropertyEntranceComponent::PropertyEntranceComponent(uint32_t componentID, Entit
 
 void PropertyEntranceComponent::OnUse(Entity* entity)
 {
+	auto* characterComponent = entity->GetComponent<CharacterComponent>();
+	auto* inventoryComponent = entity->GetComponent<InventoryComponent>();
+
+	if (!characterComponent || !inventoryComponent) return;
+
+	// Make sure we have a valid rocket
+	auto* rocket = inventoryComponent->FindItemById(characterComponent->GetLastRocketItemID());
```
Comment (EmosewaMC) [inline dGame/dComponents/PropertyEntranceComponent.cpp:34]: what if the player has binned all their rockets and made a new one?  Will this default to finding a new rocket?

File: dGame/dComponents/RocketLaunchLupComponent.cpp [hunk 4]
```diff
@@ -21,20 +21,17 @@ RocketLaunchLupComponent::RocketLaunchLupComponent(Entity* parent) : Component(p
...
-	auto* inventoryComponent = originator->GetComponent<InventoryComponent>();
 	auto* characterComponent = originator->GetComponent<CharacterComponent>();
+	auto* inventoryComponent = originator->GetComponent<InventoryComponent>();
 
-	if (!inventoryComponent || !characterComponent) return;
+	if (!characterComponent || !inventoryComponent) return;
 
-	Item* rocket = inventoryComponent->FindItemById(characterComponent->GetLastRocketItemID());
+	// Make sure we have a valid rocket
+	auto* rocket = inventoryComponent->FindItemById(characterComponent->GetLastRocketItemID());
```
Comment (EmosewaMC) [inline dGame/dComponents/RocketLaunchLupComponent.cpp:30]: Same here as above - What will happen here if the player has no previously used rocket item id?

Comment (EmosewaMC) [reply dGame/dComponents/RocketLaunchLupComponent.cpp:30]: Isnt this also supposed to equip the rocket as it does in live?  It looks like rocket isnt even used here

**Comment by aronwk-aaron** [comment]
Did not test how this works with other people till now. 
Other players cannot see you holding your rocket, so i need to fix that

File: dGame/dComponents/RocketLaunchpadControlComponent.cpp [hunk 5]
```diff
@@ -40,101 +41,36 @@ RocketLaunchpadControlComponent::~RocketLaunchpadControlComponent() {
...
-		Game::logger->Log("RocketLaunchpadControlComponent", "Unable to find rocket (%llu)!\n", optionalRocketID);
-
-		return;
-	}
+	if (!characterComponent || !inventoryComponent || !character) return;
 
-	if (rocket->GetConfig().empty()) // Sanity check
-	{
-		rocket->SetCount(0, false, false);
+	auto* rocket = inventoryComponent->FindItemById(characterComponent->GetLastRocketItemID());
```
Comment (EmosewaMC) [inline dGame/dComponents/RocketLaunchpadControlComponent.cpp:60]: same here

File: dGame/dComponents/RocketLaunchpadControlComponent.cpp [hunk 6]
```diff
@@ -143,22 +79,20 @@ void RocketLaunchpadControlComponent::Launch(Entity* originator, LWOOBJID option
...
 
 void RocketLaunchpadControlComponent::OnUse(Entity* originator) {
-	auto* propertyEntrance = m_Parent->GetComponent<PropertyEntranceComponent>();
+	auto* rocket = originator->GetComponent<CharacterComponent>()->RocketEquip(originator);
+	if (!rocket) return;
 
-	if (propertyEntrance != nullptr)
-	{
-		propertyEntrance->OnUse(originator);
+	auto* propertyEntrance = m_Parent->GetComponent<PropertyEntranceComponent>();
```
Comment (EmosewaMC) [inline dGame/dComponents/RocketLaunchpadControlComponent.cpp:91]: Would put a comment here how this returns early because the launch is handled within their respective components.

**Review by EmosewaMC — CHANGES_REQUESTED** [review]
Overall well done - few comments and questions about some bounds checks and edge cases

File: dGame/dComponents/CharacterComponent.cpp (hunk 2 above)
Comment (aronwk-aaron) [reply dGame/dComponents/CharacterComponent.cpp:473]: Copy/paste from existing code, will fix

File: dGame/dComponents/PropertyEntranceComponent.cpp (hunk 3 above)
Comment (aronwk-aaron) [reply dGame/dComponents/PropertyEntranceComponent.cpp:34]: yes, the RocketEquip function, which gets called when using the launch component will set it.
However, I think this can be done, smarter:
Remove RocketEquip from LaunchpadComp::OnUse
Add the RocketEquip to the OnUse for LUPComp and PropertyEntranceComp, and LaunchpadComp::Launch

What I was trying to do was double equip, but the ItemComp has sanity check on Equip, so it wont do that

File: dGame/dComponents/RocketLaunchLupComponent.cpp (hunk 4 above)
Comment (aronwk-aaron) [reply dGame/dComponents/RocketLaunchLupComponent.cpp:30]: the equip is handle in RocketLaunchpadComponent::OnUse, but I think the better way to do it is each component when it's actually needed, instead of doing multiple checks

**Comment by aronwk-aaron** [comment]
> Did not test how this works with other people till now. Other players cannot see you holding your rocket, so i need to fix that

TL;DR The part that made the rocket show for other players was not be serialized

**Comment by aronwk-aaron** [comment]
> Overall well done - few comments and questions about some bounds checks and edge cases

I was thinking more about this and I have a better way to do it.

> Implement proper rocket holding at launchpads.
> 
> Tested all world launchers with and without rocket. Tested all property launchers with and without rocket. Tested LUP launcher with and without rocket. Tested double clicking or dragging rocket on to each type of launcher as well and interacting with them through .
> 
> In the scenarios above, when the player has a rocket, the launcher works as expected. World launcher launch immediately Prop and LUP launchers open then menu and waits, then launch when a selection is made.
> 
> When the player has no rocket, no launch sequence happen and no menus are opened.
> 
> Tested quitting the game/killing the server while player is holding rocket. Upon next login, player is not stuck holding rocket.

Retested everything and it all worked :D
Please let me know if anything else needs to be tested

File: dGame/dComponents/CharacterComponent.cpp [hunk 7]
```diff
@@ -448,6 +449,56 @@ void CharacterComponent::SetLastRocketConfig(std::u16string config) {
...
+	}
+	return rocket;
+}
+
+Item* CharacterComponent::RocketEquip(Entity* player) {
+	Item* rocket = GetRocket(player);
+	if (!rocket) return rocket;
+
+	// build and define the rocket config
+		for (LDFBaseData* data : rocket->GetConfig()) {
```
Comment (EmosewaMC) [inline dGame/dComponents/CharacterComponent.cpp:480]: Some odd indenting here

File: dGame/dComponents/InventoryComponent.cpp [hunk 8]
```diff
@@ -797,7 +797,26 @@ void InventoryComponent::Serialize(RakNet::BitStream* outBitStream, const bool b
             
             outBitStream->Write0();
             
-            outBitStream->Write0(); //TODO: This is supposed to be true and write the assemblyPartLOTs when they're present.
+			bool flag = !item.config.empty();
```
Comment (EmosewaMC) [inline dGame/dComponents/InventoryComponent.cpp:800]: More odd indenting here

File: dGame/dComponents/PropertyEntranceComponent.cpp [hunk 9]
```diff
@@ -22,8 +23,10 @@ PropertyEntranceComponent::PropertyEntranceComponent(uint32_t componentID, Entit
     this->m_PropertyName = entry.propertyName;
 }
 
-void PropertyEntranceComponent::OnUse(Entity* entity)
-{
+void PropertyEntranceComponent::OnUse(Entity* entity) {
+	auto* rocket = entity->GetComponent<CharacterComponent>()->RocketEquip(entity);
```
Comment (EmosewaMC) [inline dGame/dComponents/PropertyEntranceComponent.cpp:27]: character component needs to be checked for nullptr here

File: dWorldServer/WorldServer.cpp [hunk 10]
```diff
@@ -972,7 +972,7 @@ void HandlePacket(Packet* packet) {
 					
                     EntityManager::Instance()->ConstructAllEntities(packet->systemAddress);
 
-					player->GetComponent<CharacterComponent>()->SetLastRocketConfig(u"");
+					player->GetComponent<CharacterComponent>()->RocketUnEquip(player);
```
Comment (EmosewaMC) [inline dWorldServer/WorldServer.cpp:975]: should also check for nullptr here as well.

**Review by EmosewaMC — CHANGES_REQUESTED** [review]
just a few spots that stood out to me that should have checks and some odd indenting.

**Comment by EmosewaMC** [comment]
What are the tests you have done for the changes?

**Comment by aronwk-aaron** [comment]
> What are the tests you have done for the changes?

I tested the all scenarios listed in the original comment again.
Also to note that I tested in GM 0 and checked racing to, since afaik, the only other item that has a config is cars

**Review by EmosewaMC — APPROVED** [review]
